Add new eddy current probe support with tap functionality

vvuk opened the pull request:
(Copied from Klipper https://github.com/Klipper3d/klipper/pull/6785 and adjusted for Kalico)

[Extensive documentation is available here](https://hackmd.io/@vvuk/r1uHAMVDkx)

Original notes from Klipper PR:

I've been working on updated eddy current sensor support for a week or so now, with a goal of supporting both homing and accurate z-offset setting via "tap" ("touch" etc.) like functionality. This works, though I'm still gathering testers (my sample size is 2 currently of people it works for, including myself). I'm opening this draft PR to get early feedback.

This is implemented as separate "ng" versions of ldc1612 code and eddy current sensor code. I ended up rewriting large parts of these, and I wanted to make sure that the current eddy current support can continue to be used as-is to make it easy for people to switch back and forth (especially during development).

Homing is similar to current `probe_eddy_current`. However, as there's no temperature calibration or compensation, homing is "coarse" at print temps. Tap is used to set an accurate physical offset at print temperatures.

"Tap" is based on an accumulated sum of a simple average of the derivative of a windowed moving average of the sensor frequencies (say that 10 times fast). I spent a while with a Jupyter notebook and a bunch of data, and this seems to give a reliable signal. There's also now an "alternate" detection implemented that processes the samples (post-MCU tap detection) through a Butterworth filter which gives even more accurate results. At some point I'm going to implement this on the MCU; however, some might not be powerful enough to do the math fast enough. So for now, the tap movement/detection has to happen on the MCU first.

Additional Kalico specific notes:

I've removed the sensor_ldc1612 and probe_eddy_current code from Kalico, as probe_eddy_current requires the new probe interface that Kalico isn't taking, and this is intended to be a replacement. Also, this supports BTT Eddy, Cartographer, and the Mellow Fly (and other ldc1612 probes, though those the ones people have done testing on so far).

Note: I haven't fully tested this on Kalico yet, beyond builds/etc. I'm out of town this week and wanted to open this draft PR to get any broad feedback (integration, approach, code style etc.) so that I can get the full PR ready next week. Any suggestions/guidance on how to do tests for this also appreciated, I'm not familiar with Klipper tests.

This is split into 3 commits; the first one expands out stm32f0x2 support for hard pwm, based on Cartographer; Carto uses this to generate a 24MHz signal to use as the reference ldc1612 signal.

- **stm32: add hard pwm for stm32f0x2, and ability to set max_pwm**
- **Add sensor_ldc161_ng**
- **Add Klippy probe_eddy_ng/ldc1612_ng**

## Checklist

- [x] pr title makes sense
- [ ] squashed to 1 commit (left as 3 on purpose, can squash in the future)
- [ ] added a test case if possible
- [x] if new feature, added to the readme
- [ ] ci is happy and green

vvuk commented:
Mmm, good ol' STM32F042x6. Given the super constrained code size there (32kb!), I don't think it's reasonable to expect that firmware can be built with every feature/support for every sensor/etc enabled for this chip. (Also, unless Kalico does this differently, this actually overflowed long ago if katapult is used.)

vvuk commented:
(I just noticed #534, so perhaps this is timely :))

rogerlz commented:
> (I just noticed #534, so perhaps this is timely :))
Indeed it is. 

> Mmm, good ol' STM32F042x6.
Yeah, it is a small one. I'd say to turn it off for it with `CONFIG_WANT_LDC1612=n` in `test/configs/stm32f042.config`

vvuk commented:
> Yeah, it is a small one. I'd say to disable ldc1612 for that mcu with CONFIG_WANT_LDC1612=n in test/configs/stm32f042.config

I can do it for the test but I'd actually like it to be tested, because Cartographer is a 32f042. I guess I can make an additional test config?

vvuk commented:
Okay, this has now been tested on my test printer; homing, probing, tap, bed mesh all seem to work.

This code is very chatty in the log. I'll put a bunch of this behind a debug config value.

I'd like to (in a separate PR) bring over some version of "bed scan"; any reason for that not to happen?

vvuk commented:
I just opened #560 to help with codesize on stm32f042; this particular 540 byte overflow is because of the hard-pwm code I enabled/added because ldc1612 is disabled in that test config. However, my klipper fits when compiled locally (normal -O2 even), with gcc `Arm GNU Toolchain 13.3.Rel1 (Build arm-13.24)) 13.3.1 20240614`.

So I think options to fix this test case are:
- Land #560 and disable something else (e.g. I2C support) to get everything to fit
- Put the hard-pwm stuff behind a WANT flag
- Upgrade gcc

recrudesce commented:
Holdup, would this replace the firmware in the Cartographer as well ?

vvuk commented:
> Holdup, would this replace the firmware in the Cartographer as well ?

It _can_, i.e. it supports that hardware. But it's not required to, of course the normal Cartographer firmware continues to work independent of this.

vvuk commented:
Updated with a bunch of additional eddy-ng work. Also added a new `stm32f042-ldc1612.config` test config that tests the small opt-for-size build.

Schmudus commented:
Very bad that eddy rapid_scan is still ot working. Is there any timeline from devs when it will be done? Can I change some files in klippy to test and use it?

vvuk commented:
Updated and added docs. The docs are much more bare-bones compared to [what's in the eddy-ng repo](https://github.com/vvuk/eddy-ng/wiki), which is probably too verbose. I'm going to keep maintaining the `eddy-ng` repo (for Klipper users), so maybe I'll add a link to those docs for a deeper dive?

vvuk commented:
Closing this after discussion with @rogerlz -- there's still a lot of development going on here, and it makes more sense to keep it as a separate plugin/module. Upstream is at https://github.com/vvuk/eddy-ng . I'll be working on making it easier to add kalico addons that have firmware modules in the next week or so to streamline using the external repo.
